### PR TITLE
Expose IsRef from IReturnOperations

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @333fred to see if i'm doing this right and ok to do.  If so, i can add tests.

Note: we need this in some of our IDE features, and we're currently going back to syntax to figure this out.  The information is in teh bound nodes, so it seems sensible that it would be in the iops.  I imagine i might be able ot figure this out in another way from iop, but it would likely take trying to map from the IReturnOperation to the containing method somehow (i.e. local func, lambda, method), and then trying to infer from that.  That feels like refiguring out what the compiler already knows.  so this approach seems preferable.

**Comment by sharwell** [comment]
> and we're currently going back to syntax to figure this out

There is no need for the IDE to go back to syntax, as demonstrated by #36272. The `OperationAnalysisContext` already contains all the information necessary to derive the `RefKind` without relying on syntax nodes.

**Comment by alrz** [comment]
Should this be really on the IReturnOperation? Woudn't FieldReference.RefKind and others make more sense?

**Comment by CyrusNajmabadi** [comment]
> already contains all the information necessary to derive the RefKind without relying on syntax nodes

I mentioned above that this seems like trying to rederive information the compiler already has.  If it's totally trivial to figure this out, that's fine.   However, it didn't seem like the way to do this was at all obvious:

> IsRef is not sufficient for IReturnOperation. The property should be RefKind, based on #36271.

I'm fine with using ref-kind.  Though i can't find anything in the language indicating that there is any sort of different refkinds for return-statements otehr than just 'by-ref'.  RefKind would open us in the future to expanding that, but i don't know if it's truly necessary.

**Comment by CyrusNajmabadi** [comment]
> Should this be really on the IReturnOperation? Woudn't FieldReference.RefKind and others make more sense?

From the language's perspective it's a 'ref-return', so i would consider it to make sense to go here.

> I see it comes from BoundReturnStatement, but what about other places where ref could appear?

What about them? :)   Things like ISimpleAssignmentOp and IConditionalOp already expose IsRef for this purpose.  These are the places the language specifies that thre are new ref syntax/semantics.  we've got it for two of the three cases, so i'm not sure why we wouldn't have it for the last place we added as well.

**Comment by CyrusNajmabadi** [comment]
So https://github.com/dotnet/roslyn/pull/36272 indicates to me exactly why i just want this property.  It's exactly as i mention in the OP.  I don't want to have to know the surrounding operation context to examine to try to reinfer this information.

**Comment by sharwell** [comment]
I'm not saying it's not a valuable addition. I'm just pointing out that it's not fundamentally required as part of avoiding syntax operations to determine it.

**Comment by CyrusNajmabadi** [comment]
> I'm not saying it's not a valuable addition. I'm just pointing out that it's not fundamentally required as part of avoiding syntax operations to determine it.

Right.. but i was making two points originally, that you need to go back to syntax *or* you need to reinfer what the compiler already knows here.  Neither are particularly palatable here.  :)

**Comment by sharwell** [comment]
Note that Make Field Readonly cannot be implemented using `IsRef` alone. The analyzer needs to distinguish between `ref` and `ref readonly` (the latter is implied by the return type).

**Comment by CyrusNajmabadi** [comment]
Another, complimentary, options for me would be if we could easily expose the method symbol that the IOp is returning from.  That would make it far more pleasant to go and figure out waht sort of ref-type this return op was.

**Comment by 333fred** [comment]
We definitely would want to use `RefKind` and not a bool flag.

I also wouldn't be opposed to having a `ContainingSymbol` property on the appropriate contexts to expose the current symbol. @AlekseyTs or @gafter, do you know why such a property isn't exposed today already?

**Comment by gafter** [comment]
@333fred I think we just assumed that whomever is visiting the IOperation nodes already knows the method from which they got them, but I have no objection to making the method available from the "return" operation.

**Comment by sharwell** [comment]
@gafter the assumption about knowing the containing symbol falls apart for lambdas, anonymous functions, and local functions. The operation context only gives the outer method symbol.

**Comment by AlekseyTs** [comment]
> the assumption about knowing the containing symbol falls apart for lambdas, anonymous functions, and local functions.

I think SemanticModel has GetEnclosingSymbol API. I also do not like exposing this just for IReturnOperation.

**Comment by 333fred** [comment]
http://sourceroslyn.io/#Microsoft.CodeAnalysis/Compilation/SemanticModel.cs,7fd8f5c1c767ad27

@CyrusNajmabadi, would that API satisfy your need there? I would probably be fine with adding an overload that takes an IOperation for convenience if needed.

**Comment by CyrusNajmabadi** [comment]
My concern with that api is that it's just doc'ed vaguely as:

>        /// <summary>
>        /// Given a position in the SyntaxTree for this SemanticModel returns the innermost Symbol
>        /// that the position is considered inside of.
>        /// </summary>

It's not really clear what that means, or if it might break in the future (say with the addition of some new symbol).  We're really trying to ask "which method is this returning from". 

>  I would probably be fine with adding an overload that takes an IOperation for convenience if needed.

 If we have a clear helper for that purpose, then we know what semantics it must have even as we change the lang/compiler-apis.  I would be totally ok with that.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -3560,7 +3560,7 @@ public override IOperation VisitReturn(IReturnOperation operation, int? captureI
             switch (operation.Kind)
             {
                 case OperationKind.YieldReturn:
-                    AddStatement(new ReturnOperation(returnedValue, OperationKind.YieldReturn, semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, IsImplicit(operation)));
+                    AddStatement(new ReturnOperation(returnedValue, isRef: false, OperationKind.YieldReturn, semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, IsImplicit(operation)));
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3563]: >isRef: false [](start = 68, length = 12)

Would it be more robust to use value from the original node?

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml [hunk 2]
```diff
@@ -389,6 +389,11 @@
         <summary>Value to be returned.</summary>
       </Comments>
     </Property>
+    <Property Name="IsRef" Type="bool">
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/OperationInterfaces.xml:392]: >IsRef [](start = 20, length = 5)

RefKind?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 3]
```diff
@@ -1648,7 +1648,8 @@ private IReturnOperation CreateBoundReturnStatementOperation(BoundReturnStatemen
             ITypeSymbol type = null;
             Optional<object> constantValue = default(Optional<object>);
             bool isImplicit = boundReturnStatement.WasCompilerGenerated;
-            return new CSharpLazyReturnOperation(this, returnedValue, OperationKind.Return, _semanticModel, syntax, type, constantValue, isImplicit);
+            bool isRef = boundReturnStatement.RefKind == RefKind.Ref;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1651]: >boundReturnStatement.RefKind [](start = 25, length = 28)

Could this be ref-readonly?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1).
I would expect some existing IOperation tree tests be affected by the change, or new tests added to fill the gap.
I would also expect two test IOperation visitors that we have to be adjusted to handle the new property.

**Comment by CyrusNajmabadi** [comment]
Closing out.

**Pull request closed without merging**